ctags/parsers/autoit.c

```diff
@@ -0,0 +1,121 @@
+/*
+*   Copyright (c) 2017, Alexey Olenich
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License version 2 or (at your option) any later version.
+*
+*   This module contains functions for generating tags for AutoIt functions.
+*   Homepage             https://www.autoitscript.com/site/autoit/
+*   Online Documentation https://www.autoitscript.com/autoit3/docs/
+*
+*   Functions: (?i)^func[ \t]{1,}([a-z0-9_]{1,}) >> \1
+*   Regions:   (?i)^#region[ \t]*(.*?)           >> \1
+*/
+
+/*
+*   INCLUDE FILES
+*/
+#include "general.h"  /* must always come first */
+
+#include <string.h>
+
+#include "parse.h"
+#include "read.h"
+#include "routines.h"
+#include "vstring.h"
+
+/*
+*   DATA DEFINITIONS
+*/
+typedef enum {
+	K_FUNCTION,
+	K_REGION
+} AutoItKind;
+
+static kindOption AutoItKinds [] = {
+	{ true, 'f', "func", "functions" },
+	{ true, 'r', "region", "regions" }
+};
+
+/*
+*   FUNCTION DEFINITIONS
+*/
+static void findAutoItTags (void)
+{
+	vString *name = vStringNew ();
+	const unsigned char *line;
+
+	while ((line = readLineFromInputFile ()) != NULL)
+	{
+		const unsigned char* p = line;
+		if (p [0] == '#')
+		{
+			/* min. string "#region" > 7 */
+			if ((p [1] == 'R' || p [1] == 'r') &&
+				strlen ((const char *) p) > 8 &&
```

Comment:
How can I use subscript (for examle ```p [1]```) if I don't know subscript dimension range, i.e. string length?

Reply:
C strings have a last element with value `0`, which is actually delimiting the end of the string. `strlen(str)` is just going through `str` to find at which offset `0` is found (it could basically be implemented as `for (len = 0; str[len] != 0; len++);`).

Anyway, this means that so long as none of the characters before the subscript has value `0`, the next one is a valid subscript -- it just can be the last one if it has value `0`.  Thus, as you check the values of subscripts successively against a defined set *not including* `0`, you're sure that you won't try subscripting past the end of the string, otherwise your tests would have to be false already anyway.
So if `p[1]` is either `'R'` or `'r'` (so not `0`), `p[2]` has to be a valid subscript.

Reply:
I remembered why I used ```strlen()``` :) Besides subscript.
Function defining:
```autoit
Func t()
```
function name is ```t``` and one symbol is a minimum length of name. If string beginning with ```func``` then we have define of function , but if string length less 8 symbols then we have not valid function name, i.e. we have not function name, and this string should be discarded.

Reply:
I still think it's irrelevant, because it's then an "optimization" (as the code after more correctly checks the syntax), but it only kicks in when the there is less than an extra 8 characters after the line starting with `f`. This is highly unlikely to result in any speed benefits (especially as you're checking the next characters for `unc[ \t]`), and is highly likely to let pass things that aren't a function anyway but just happen to be long enough.

In any case, unless profiling shows this is faster (I doubt it), the clearest/simplest thing should be used IMO, which IMO again, is dropping this check.